chain/epoch_manager/src/lib.rs

```diff
@@ -1464,73 +1440,151 @@ impl EpochManager {
-    /// from cache and invalidates the cache.
-    pub fn get_and_update_epoch_info_aggregator(
+    /// Updates epoch info aggregator to state as of `last_final_block_hash`
+    /// block.
+    ///
+    /// The block hash passed as argument should be a final block so that the
+    /// method can perform efficient incremental updates.  Calling this method
+    /// on a block which has not been finalised yet is likely to result in
+    /// performance issues since handling forks will force it to traverse the
+    /// entire epoch from scratch.
+    ///
+    /// The result of the aggregation is stored in `self.epoch_info_aggregator`.
+    ///
+    /// Saves the aggregator to `store_update` if epoch id changes or every
+    /// AGGREGATOR_SAVE_PERIOD heights.
+    pub fn update_epoch_info_aggregator_upto_final(
+        &mut self,
+        last_final_block_hash: &CryptoHash,
+        store_update: &mut StoreUpdate,
+    ) -> Result<(), EpochError> {
+        if let Some((aggregator, replace)) =
+            self.aggregate_epoch_info_upto(last_final_block_hash)?
+        {
+            let save = if replace {
+                self.epoch_info_aggregator = aggregator;
+                true
+            } else {
+                self.epoch_info_aggregator.merge(aggregator);
+                let block_info = self.get_block_info(last_final_block_hash)?;
+                *block_info.height() % AGGREGATOR_SAVE_PERIOD == 0
+            };
+            if save {
+                store_update.set_ser(ColEpochInfo, AGGREGATOR_KEY, &self.epoch_info_aggregator)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Returns epoch info aggregate with state up to `last_block_hash`.
+    ///
+    /// The block hash passed as argument should be the latest block belonging
+    /// to current epoch.  Calling this method on any other block is likely to
+    /// result in performance issues since handling something which is not past
+    /// the final block will force it to traverse the entire epoch from scratch.
+    ///
+    /// This method does not change `self.epoch_info_aggregator`.
+    pub fn get_epoch_info_aggregator_upto_last(
         &mut self,
-        epoch_id: &EpochId,
         last_block_hash: &CryptoHash,
-        copy_only: bool,
     ) -> Result<EpochInfoAggregator, EpochError> {
-        let epoch_info_aggregator_cache = if copy_only {
-            self.epoch_info_aggregator.clone()
-        } else {
-            self.epoch_info_aggregator.take()
-        };
-        let mut epoch_change = false;
-        let mut aggregator = if let Some(aggregator) = epoch_info_aggregator_cache {
-            aggregator
+        if let Some((mut aggregator, replace)) = self.aggregate_epoch_info_upto(last_block_hash)? {
+            if !replace {
+                aggregator.merge_prefix(&self.epoch_info_aggregator);
+                aggregator.last_block_hash = *last_block_hash;
```

Comment:
I'm surprised that you have to set it here. Is 'merge_prefix' overwriting it?

Reply:
Good catch. I meant to remove this once I introduced `merge_prefix`.